File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKNode.java
package br.com.insanitech.spritekit;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Dictionary;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

import br.com.insanitech.spritekit.opengl.renderer.GLRenderer;

public class SKNode implements GLRenderer.GLDrawer {
    public interface ChildNodesEnumeration {
        void nextChildNode(SKNode node, boolean shouldStop);
    }

    private SKRect frame = new SKRect(0.0f, 0.0f, 0.0f, 0.0f);
    private SKNode parent;
    private String name = "";

    private SKScene scene;

    private SKPhysicsBody physicsBody;
    private SKReachConstraints reachConstraints;
    private List<SKReachConstraints> constraints;
    private Dictionary<?, ?> userData;

    private final List<SKNode> children = new ArrayList<SKNode>();
    private final LinkedList<SKAction> actions = new LinkedList<SKAction>();

    final SKPoint position = new SKPoint(0.0f, 0.0f);
    boolean paused = false;
    boolean hidden = false;
    float zPosition = 0.0f;
    float zRotation = 0.0f;
    float xScale = 1.0f;
    float yScale = 1.0f;
    float speed = 1.0f;
    float alpha = 1;
    boolean userInteractionEnabled = true;

    public static SKNode node() {
        return new SKNode();
    }

    public SKPoint getPosition() {
        synchronized (this) {
            return position;
        }
    }

    public void setPosition(SKPoint position) {
        synchronized (this) {
            this.position.assignByValue(position);
        }
    }

    public void setPosition(float x, float y) {
        synchronized (this) {
            position.x = x;
            position.y = y;
        }
    }

    public boolean isPaused() {
        synchronized (this) {
            return paused;
        }
    }

    public void setPaused(boolean paused) {
        synchronized (this) {
            this.paused = paused;
        }
    }

    public boolean isHidden() {
        synchronized (this) {
            return hidden;
        }
    }

    public void setHidden(boolean hidden) {
        synchronized (this) {
            this.hidden = hidden;
        }
    }

    public float getZPosition() {
        synchronized (this) {
            return zPosition;
        }
    }

    public void setZPosition(float zPosition) {
        synchronized (this) {
            this.zPosition = zPosition;
        }
    }

    public float getZRotation() {
        synchronized (this) {
            return zRotation;
        }
    }

    public void setZRotation(float zRotation) {
        synchronized (this) {
            this.zRotation = zRotation;
        }
    }

    public float getScaleX() {
        synchronized (this) {
            return xScale;
        }
    }

    public void setScaleX(float xScale) {
        synchronized (this) {
            this.xScale = xScale;
        }
    }

    public float getScaleY() {
        synchronized (this) {
            return yScale;
        }
    }

    public void setScaleY(float yScale) {
        synchronized (this) {
            this.yScale = yScale;
        }
    }

    public void setScale(float scale) {
        synchronized (this) {
            xScale = yScale = scale;
        }
    }

    public float getSpeed() {
        synchronized (this) {
            return speed;
        }
    }

    public void setSpeed(float speed) {
        synchronized (this) {
            this.speed = speed;
        }
    }

    public float getAlpha() {
        synchronized (this) {
            return alpha;
        }
    }

    public void setAlpha(float alpha) {
        synchronized (this) {
            this.alpha = alpha;
        }
    }

    public boolean isUserInteractionEnabled() {
        synchronized (this) {
            return userInteractionEnabled;
        }
    }

    public void setUserInteractionEnabled(boolean userInteractionEnabled) {
        synchronized (this) {
            this.userInteractionEnabled = userInteractionEnabled;
        }
    }

    @Override
    public void onDrawFrame(GLRenderer renderer, int width, int height) {
        synchronized (this) {
            renderer.translate(position.x, position.y, zPosition);

            renderer.rotate(0, 0, zRotation);
            renderer.scale(xScale, yScale);

            drawChildren(renderer, width, height);
        }
    }

    void drawChildren(GLRenderer renderer, int width, int height) {
        try {
            List<SKNode> children = new ArrayList<SKNode>(this.children);
            for (int i = 0; i < children.size(); i++) {
                renderer.saveState();
                children.get(i).onDrawFrame(renderer, width, height);
                renderer.restoreState();
            }
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    void evaluateActions() {
        synchronized (this) {
            if (!paused) {
                List<SKAction> actions = new ArrayList<SKAction>(this.actions);
                for (SKAction action : actions) {
                    action.start();
                    action.computeAction();
                }
            }

            List<SKNode> children = new ArrayList<SKNode>(this.children);
            for (int i = 0; i < children.size(); i++) {
                children.get(i).evaluateActions();
            }
        }
    }

    public SKRect calculateAccumulatedFrame() {
        synchronized (this) {
            return null;
        }
    }

    private void setSceneRecursive(SKNode current, SKScene scene) {
        List<SKNode> children = new ArrayList<SKNode>(this.children);
        for (int i = 0; i < children.size(); i++) {
            current.setSceneRecursive(children.get(i), scene);
        }
        current.scene = scene;
    }

    public void addChild(SKNode node) {
        synchronized (this) {
            children.add(node);
            node.parent = this;
            setSceneRecursive(node, scene);
        }
    }

    public void insertChildAt(SKNode node, int index) {
        synchronized (this) {
            children.add(index, node);
            node.parent = this;
            setSceneRecursive(node, scene);
        }
    }

    public void removeChildren(List<SKNode> children) {
        synchronized (this) {
            this.children.removeAll(children);
            SKNode child;
            for (int i = 0; i < children.size(); i++) {
                child = children.get(i);
                child.parent = null;
                setSceneRecursive(child, null);
            }
        }
    }

    public void removeAllChildren() {
        synchronized (this) {
            List<SKNode> children = new ArrayList<SKNode>(this.children);
            SKNode child;
            for (int i = 0; i < children.size(); i++) {
                child = children.get(i);
                child.parent = null;
                setSceneRecursive(child, null);
            }
            this.children.clear();
        }
    }

    public void removeFromParent() {
        synchronized (this) {
            if (parent != null) {
                parent.removeChildren(Collections.singletonList(this));
                parent = null;
                setSceneRecursive(this, null);
            }
        }
    }

    public SKNode childNode(String name) {
        synchronized (this) {
            List<SKNode> children = new ArrayList<SKNode>(this.children);
            SKNode child;
            for (int i = 0; i < children.size(); i++) {
                child = children.get(i);
                if (name.equals(child.name)) {
                    return child;
                }
            }
            return null;
        }
    }

    public void enumerateChildNodes(String name, ChildNodesEnumeration enumeration) throws Exception {
        synchronized (this) {
            throw new Exception("NotImplementedException");
        }
    }

    public boolean inParentHierarchy(SKNode parent) {
        synchronized (this) {
            SKNode current = this;
            while (current.parent != null) {
                current = current.parent;
                if (current.parent == parent) {
                    return true;
                }
            }
            return false;
        }
    }

    void actionCompleted(SKAction completed) {
        synchronized (this) {
            actions.remove(completed);
        }
    }

    public void runAction(SKAction action) {
        synchronized (this) {
            action.setParent(this);
            Random rand = new Random();
            action.key = rand.nextInt() + "none" + rand.nextInt();
            actions.add(action);
        }
    }

    public void runAction(SKAction action, Runnable completion) {
        synchronized (this) {
            action.setParent(this);
            Random rand = new Random();
            action.key = rand.nextInt() + "none" + rand.nextInt();
            action.completion = completion;
            actions.add(action);
        }
    }

    public void runAction(SKAction action, String key) {
        synchronized (this) {
            action.setParent(this);
            action.key = key;
            actions.add(action);
        }
    }

    public boolean hasActions() {
        synchronized (this) {
            return actions.size() > 0;
        }
    }

    public SKAction getAction(String key) {
        synchronized (this) {
            List<SKAction> actions = new ArrayList<SKAction>(this.actions);
            for (SKAction action : actions) {
                if (action.key.equals(key)) {
                    return action;
                }
            }
        }
        return null;
    }

    public void removeAction(String key) {
        synchronized (this.actions) {
            List<SKAction> actions = new ArrayList<SKAction>(this.actions);
            for (SKAction action : actions) {
                if (action.key.equals(key)) {
                    synchronized (this.actions) {
                        this.actions.remove(action);
                    }
                }
            }
        }
    }

    public void removeAllActions() {
        synchronized (this) {
            actions.clear();
        }
    }

    public boolean containsPoint(SKPoint p) {
        synchronized (this) {
            return getFrame().containsPoint(p);
        }
    }

    public SKNode nodeAt(SKPoint p) {
        synchronized (this) {
            // TODO: implement nodeAt
            return null;
        }
    }

    public SKNode[] nodesAt(SKPoint p) {
        synchronized (this) {
            // TODO: implement nodesAt
            return null;
        }
    }

    public SKPoint convertFrom(SKPoint p, SKNode node) {
        synchronized (this) {
            // TODO: implement convertFrom
            return null;
        }
    }

    public SKPoint convertTo(SKPoint p, SKNode node) {
        synchronized (this) {
            // TODO: implement convertTo
            return null;
        }
    }

    public boolean intersects(SKNode node) {
        synchronized (this) {
            // TODO: implement intersects
            return false;
        }
    }

    public String getName() {
        synchronized (this) {
            return name;
        }
    }

    public void setName(String name) {
        synchronized (this) {
            this.name = name;
        }
    }

    public Dictionary<?, ?> getUserData() {
        synchronized (this) {
            return userData;
        }
    }

    public void setUserData(Dictionary<?, ?> userData) {
        synchronized (this) {
            this.userData = userData;
        }
    }

    public SKRect getFrame() {
        synchronized (this) {
            // TODO: implement frame to contain content bounds
            return frame;
        }
    }

    public SKNode getParent() {
        synchronized (this) {
            return parent;
        }
    }

    public List<SKNode> getChildren() {
        synchronized (this) {
            return children;
        }
    }

    public SKScene getScene() {
        synchronized (this) {
            return scene;
        }
    }

    public SKPhysicsBody getPhysicsBody() {
        synchronized (this) {
            return physicsBody;
        }
    }

    public void setPhysicsBody(SKPhysicsBody physicsBody) {
        synchronized (this) {
            this.physicsBody = physicsBody;
        }
    }

    public SKReachConstraints getReachConstraints() {
        synchronized (this) {
            return reachConstraints;
        }
    }

    public void setReachConstraints(SKReachConstraints reachConstraints) {
        synchronized (this) {
            this.reachConstraints = reachConstraints;
        }
    }

    public List<SKReachConstraints> getConstraints() {
        synchronized (this) {
            return constraints;
        }
    }

    public void setConstraints(List<SKReachConstraints> constraints) {
        synchronized (this) {
            this.constraints = constraints;
        }
    }

    protected SKNode copy(SKNode input) {
        input.frame.assignByValue(frame);
        input.parent = parent;
        input.name = name;

        input.scene = scene;

        input.physicsBody = physicsBody;
        input.reachConstraints = reachConstraints;
        input.constraints = constraints;
        input.userData = userData;

        input.children.addAll(children);
        input.actions.addAll(actions);

        input.position.assignByValue(position);
        input.paused = paused;
        input.hidden = hidden;
        input.zPosition = zPosition;
        input.zRotation = zRotation;
        input.xScale = xScale;
        input.yScale = yScale;
        input.speed = speed;
        input.alpha = alpha;
        input.userInteractionEnabled = userInteractionEnabled;

        return input;
    }

    public Object copy() {
        synchronized (this) {
            return copy(new SKNode());
        }
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKNode.kt
package br.com.insanitech.spritekit

import br.com.insanitech.spritekit.actions.SKAction
import br.com.insanitech.spritekit.core.SKBlock
import br.com.insanitech.spritekit.graphics.SKPoint
import br.com.insanitech.spritekit.graphics.SKRect
import br.com.insanitech.spritekit.opengl.renderer.GLRenderer

open class SKNode {
    internal val actions = ArrayList<SKAction>()

    private var childrenNodes: ArrayList<SKNode> = ArrayList()

    var position: SKPoint = SKPoint()
        set(value) {
            field.point.assignByValue(value.point)
        }

    var zPosition: Float = 0.0f
    var xScale: Float = 1.0f
    var yScale: Float = 1.0f
    var zRotation: Float = 0.0f
    var alpha: Float = 1.0f
    var isHidden: Boolean = false
    var isUserInteractionEnabled: Boolean = false

    val children: List<SKNode>
        get() = this.childrenNodes

    var parent: SKNode? = null
    var scene: SKScene? = null
    var name: String? = null
    var speed: Float = 1.0f
    var isPaused: Boolean = false

    internal open val isSurface: Boolean = false

    fun setScale(scale: Float) {
        this.xScale = scale
        this.yScale = scale
    }

    fun setPosition(x: Float, y: Float) {
        this.position.x = x
        this.position.y = y
    }

    private fun isNodeAtPointCandidate(p: SKPoint, node: SKNode): Boolean {
        val basicCondition = !node.isHidden && node.alpha > 0.0f && node.isSurface
        val accumulatedFrameContainsPoint = basicCondition && node.calculateAccumulatedFrame().containsPoint(p)
        return basicCondition && accumulatedFrameContainsPoint
    }

    internal open fun convertPointForCandidateParentIfNeeded(p: SKPoint, parent: SKNode) : SKPoint = p

    private fun sort(nodes: List<SKNode>) : List<SKNode> {
        return nodes.sortedWith(Comparator { o1, o2 ->
            when {
                o1.zPosition > o2.zPosition -> return@Comparator -1
                o1.zPosition == o2.zPosition -> return@Comparator 0
                o1.zPosition < o2.zPosition -> return@Comparator 1
                else -> return@Comparator 0
            }
        })
    }

    private fun forEachAtPointCandidateChildren(p: SKPoint, node: SKNode) : List<SKNode> {
        val possibleNodes = mutableListOf<SKNode>()
        this.sort(node.childrenNodes).forEach {
            val point = this.convertPointForCandidateParentIfNeeded(p, node)
            if (this.isNodeAtPointCandidate(point, it)) possibleNodes.add(it)
            possibleNodes.addAll(this.forEachAtPointCandidateChildren(p, it))
        }
        return possibleNodes
    }

    private fun atPoint(p: SKPoint, node: SKNode) : SKNode {
        val filtered = this.forEachAtPointCandidateChildren(p, node)
        val sorted = this.sort(filtered)
        return sorted.firstOrNull() ?: this
    }

    open fun atPoint(p: SKPoint): SKNode = this.atPoint(p, this)

    open fun convertFrom(point: SKPoint, node: SKNode): SKPoint {
        var xParentRelated = 0.0f
        var yParentRelated = 0.0f
        var xNodeRelated = 0.0f
        var yNodeRelated = 0.0f

        this.forEachParent {
            xParentRelated += it.position.x
            yParentRelated += it.position.y
        }
        node.forEachParent {
            xNodeRelated += it.position.x
            yNodeRelated += it.position.y
        }

        return SKPoint(point.x - xParentRelated + xNodeRelated,
                point.y - yParentRelated + yNodeRelated)
    }

    open fun convertTo(point: SKPoint, node: SKNode): SKPoint {
        var xParentRelated = 0.0f
        var yParentRelated = 0.0f
        var xNodeRelated = 0.0f
        var yNodeRelated = 0.0f

        this.forEachParent {
            xParentRelated += it.position.x
            yParentRelated += it.position.y
        }
        node.forEachParent {
            xNodeRelated += it.position.x
            yNodeRelated += it.position.y
        }

        return SKPoint(point.x + xParentRelated + xNodeRelated,
                point.y + yParentRelated + yNodeRelated)
    }

    internal fun accumulateFrame(ofRect: SKRect, inRect: SKRect) {
        if (inRect.x > ofRect.x) {
            inRect.origin.x = ofRect.x
        }
        if (inRect.y > ofRect.y) {
            inRect.origin.y = ofRect.y
        }
        if (inRect.width == 0f) {
            inRect.size.width = ofRect.width
        } else if (inRect.x + inRect.width < ofRect.x + ofRect.width) {
            inRect.size.width += (ofRect.x + ofRect.width) - (inRect.x + inRect.width)
        }
        if (inRect.height == 0f) {
            inRect.size.height = ofRect.height
        } else if (inRect.y + inRect.height < ofRect.y + ofRect.height) {
            inRect.size.height += (ofRect.y + ofRect.height) - (inRect.y + inRect.width)
        }
    }

    internal open fun calculateAccumulatedFrame(adjustParent: Boolean): SKRect {
        val rect = SKRect()
        this.childrenNodes.forEach { this.accumulateFrame(it.calculateAccumulatedFrame(false), rect) }
        if (adjustParent && this.parent != null) {
            rect.origin = this.convertTo(rect.origin, this.parent!!)
        }
        return rect
    }

    open fun calculateAccumulatedFrame(): SKRect = this.calculateAccumulatedFrame(true)

    open fun addChild(node: SKNode) {
        this.addChild(node, true)
    }

    internal fun addChild(node: SKNode, move: Boolean) {
        this.childrenNodes.add(node)
        node.parent = this
        if (move) {
            node.movedToScene(this.scene)
        }
    }

    open fun insertChild(node: SKNode, index: Int) {
        this.insertChild(node, index, true)
    }

    internal fun insertChild(node: SKNode, index: Int, move: Boolean) {
        this.childrenNodes.add(index, node)
        node.parent = this
        if (move) {
            node.movedToScene(this.scene)
        }
    }

    open fun removeFromParent() {
        this.parent?.childrenNodes?.remove(this)
        this.parent = null
        this.movedToScene(null)
    }

    open fun removeAllChildren() {
        val children = ArrayList<SKNode>(this.childrenNodes)
        children.forEach(SKNode::removeFromParent)
    }

    open fun removeChildren(nodes: List<SKNode>) {
        val list = ArrayList<SKNode>(nodes)
        list.forEach { if (it.parent == this) it.removeFromParent() }
    }

    internal fun forEachParent(block: (parent: SKNode) -> Unit) {
        var myParent: SKNode? = this
        while (myParent != null) {
            block(myParent)
            myParent = myParent.parent
        }
    }

    fun inParentHierarchy(parent: SKNode): Boolean {
        var status = false
        this.forEachParent {
            if (it == parent) {
                status = true; return@forEachParent
            }
        }
        return status
    }

    internal open fun movedToScene(scene: SKScene?) {
        this.scene = scene
        this.childrenNodes.forEach {
            it.movedToScene(scene)
        }
    }

    fun run(action: SKAction) {
        this.actions.add(action)
    }

    fun run(action: SKAction, completion: SKBlock) {
        action.completion = completion
        this.actions.add(action)
    }

    fun run(action: SKAction, key: String) {
        this.removeAction(key)
        action.key = key
        this.actions.add(action)
    }

    fun action(key: String): SKAction? = this.actions.firstOrNull { it.key == key }

    fun hasActions() = this.actions.size > 0

    fun removeAllActions() {
        this.actions.clear()
    }

    fun removeAction(key: String) {
        val action = this.action(key)
        when {
            action != null -> this.actions.remove(action)
        }
    }

    open fun copy(): SKNode = this.copy(SKNode())

    protected fun <T : SKNode> copy(into: T): T {
        val newNode = into
        newNode.position = this.position
        newNode.alpha = this.alpha
        newNode.isHidden = this.isHidden
        newNode.isPaused = this.isPaused
        newNode.isUserInteractionEnabled = this.isUserInteractionEnabled
        newNode.name = this.name
        newNode.speed = this.speed
        newNode.xScale = this.xScale
        newNode.yScale = this.yScale
        newNode.zPosition = this.zPosition
        newNode.zRotation = this.zRotation
        return newNode
    }

    override fun toString(): String =
            "[${this.javaClass.simpleName}] name: ${this.name}, position: ${this.position}, scale: ${SKPoint(this.xScale, this.yScale)}, rotation: ${this.zRotation}"

    // MARK: Drawer implementations

    internal open val drawer: SKNodeDrawer by lazy { SKNodeDrawer(this) }

    internal open class SKNodeDrawer(protected open val node: SKNode) : GLRenderer.GLDrawer {
        override fun drawFrame(renderer: GLRenderer, width: Int, height: Int) {
            renderer.translate(this.node.position.x, this.node.position.y, this.node.zPosition)
            renderer.rotate(0f, 0f, this.node.zRotation)
            renderer.scale(this.node.xScale, this.node.yScale)
            this.drawChildren(renderer, width, height)
        }

        internal fun drawChildren(renderer: GLRenderer, width: Int, height: Int) {
            ArrayList(this.node.children).sortedBy { it.zPosition }.forEach {
                it.drawer.drawFrame(renderer, width, height)
            }
        }
    }

    companion object {
        fun node(): SKNode = SKNode()
    }
}
